#####
## submission via API for rat files (taxon id: 10116)
#
TAXON=10116
SPECIES=RGD
APITOKEN=`cat api.token`

ALLELE_SPEC="${AGR_RELEASE}_ALLELE_${SPECIES}"
ALLELE_LOC="data/$BATCH/${MOD_NAME}_${AGR_VER}_allele.${TAXON}.json"
gzip $ALLELE_LOC
ALLELE_FILE="${ALLELE_SPEC}=@${ALLELE_LOC}.gz"
echo "$ALLELE_FILE"

VARIANT_SPEC="${AGR_RELEASE}_VARIATION_${SPECIES}"
VARIANT_LOC="data/$BATCH/${MOD_NAME}_${AGR_VER}_variant.${TAXON}.json"
gzip $VARIANT_LOC
VARIANT_FILE="${VARIANT_SPEC}=@${VARIANT_LOC}.gz"
echo "$VARIANT_FILE"

EXPR_SPEC="${AGR_RELEASE}_EXPRESSION_${SPECIES}"
EXPR_LOC="data/$BATCH/${MOD_NAME}_${AGR_VER}_expression.${TAXON}.json"
gzip $EXPR_LOC
EXPRESSION_FILE="${EXPR_SPEC}=@${EXPR_LOC}.gz"
echo "$EXPRESSION_FILE"

PHENOTYPE_SPEC="${AGR_RELEASE}_PHENOTYPE_${SPECIES}"
PHENOTYPE_LOC="data/$BATCH/${MOD_NAME}_${AGR_VER}_phenotype.${TAXON}.json"
PHENOTYPE_FILE="${PHENOTYPE_SPEC}=@${PHENOTYPE_LOC}"
echo "$PHENOTYPE_FILE"

DISEASE_SPEC="${AGR_RELEASE}_DAF_${SPECIES}"
DISEASE_LOC="data/$BATCH/${MOD_NAME}_${AGR_VER}_disease.${TAXON}.json"
DISEASE_FILE="${DISEASE_SPEC}=@${DISEASE_LOC}"
echo "$DISEASE_FILE"

GENE_SPEC="${AGR_RELEASE}_BGI_${SPECIES}"
GENE_LOC="data/$BATCH/${MOD_NAME}_${AGR_VER}_gene.${TAXON}.json"
GENE_FILE="${GENE_SPEC}=@${GENE_LOC}"
echo "$GENE_FILE"

AGM_SPEC="${AGR_RELEASE}_AGM_${SPECIES}"
AGM_LOC="data/$BATCH/${MOD_NAME}_${AGR_VER}_affectedGenomicModel.${TAXON}.json"
AGM_FILE="${AGM_SPEC}=@${AGM_LOC}"
echo "$AGM_FILE"

GFF_SPEC="${AGR_RELEASE}_GFF_${SPECIES}"
GFF_LOC="data/$BATCH/${MOD_NAME}_${AGR_VER}_GFF.${TAXON}.gff3"
gzip $GFF_LOC
GFF_FILE="${GFF_SPEC}=@${GFF_LOC}.gz"
echo "$GFF_FILE"

HTPDATASET_SPEC="${AGR_RELEASE}_HTPDATASET_${SPECIES}"
HTPDATASET_LOC="data/$BATCH/${MOD_NAME}_${AGR_VER}_HTPDATASET.${TAXON}.json"
HTPDATASET_FILE="${HTPDATASET_SPEC}=@${HTPDATASET_LOC}"
echo "$HTPDATASET_FILE"

HTPDATASAMPLE_SPEC="${AGR_RELEASE}_HTPDATASAMPLE_${SPECIES}"
HTPDATASAMPLE_LOC="data/$BATCH/${MOD_NAME}_${AGR_VER}_HTPDATASAMPLE.${TAXON}.json"
HTPDATASAMPLE_FILE="${HTPDATASAMPLE_SPEC}=@${HTPDATASAMPLE_LOC}"
echo "$HTPDATASAMPLE_FILE"

curl -k \
 -H "Authorization: Bearer $APITOKEN" \
 -X POST "https://fms.alliancegenome.org/api/data/submit" \
 -F "$ALLELE_FILE" \
 -F "$VARIANT_FILE" \
 -F "$EXPRESSION_FILE" \
 -F "$PHENOTYPE_FILE" \
 -F "$DISEASE_FILE" \
 -F "$GENE_FILE" \
 -F "$GFF_FILE" \
 -F "$AGM_FILE" \
 -F "$HTPDATASET_FILE" \
 -F "$HTPDATASAMPLE_FILE" \
 | tee api_rat_submission2.log


REFERENCE_SPEC="${AGR_RELEASE}_REFERENCE_${SPECIES}"
REFERENCE_LOC="data/$BATCH/${MOD_NAME}_${AGR_VER}_REFERENCE.${TAXON}.json"
gzip $REFERENCE_LOC
REFERENCE_FILE="${REFERENCE_SPEC}=@${REFERENCE_LOC}.gz"
echo "$REFERENCE_FILE"

REFEXCHANGE_SPEC="${AGR_RELEASE}_REF-EXCHANGE_${SPECIES}"
REFEXCHANGE_LOC="data/$BATCH/${MOD_NAME}_${AGR_VER}_REF-EXCHANGE.${TAXON}.json"
REFEXCHANGE_FILE="${REFEXCHANGE_SPEC}=@${REFEXCHANGE_LOC}"
echo "$REFEXCHANGE_FILE"

curl -k \
 -H "Authorization: Bearer $APITOKEN" \
 -X POST "https://fms.alliancegenome.org/api/data/submit" \
 -F "$REFERENCE_FILE" \
 -F "$REFEXCHANGE_FILE" \
 | tee ref_submission.log

echo "=== OK ==="

